src/main/scala/io/iohk/ethereum/domain/Blockchain.scala

```diff
@@ -232,9 +243,15 @@ class BlockchainImpl(
   override protected def getHashByBlockNumber(number: BigInt): Option[ByteString] =
     blockNumberMappingStorage.get(number)
 
-  private def saveBlockNumberMapping(number: BigInt, hash: ByteString): Unit = blockNumberMappingStorage.put(number, hash)
+  private def saveBlockNumberMapping(number: BigInt, hash: ByteString): Unit = {
+    blockNumberMappingStorage.put(number, hash)
+    appStateStorage.putBestBlockNumber(number) // FIXME: mother of consistency?!?!
```

Comment:
Yeah, so I wanted avoid having yet another save (in the series of three), so I went with this. But this is terrible because it expects blocks to be saved and removed in specific order - I don't think we'll ever want to save them in a different order but it's still bad.

What if the `save`/`remove` methods took another optional param, that would determine whether the best block number should be updated and how?

(I also thought adding a bit more logic to the best number updates in those methods, but I think that would be wrong, not only because of an extra interaction with the storage)

Reply:
>What if the save/remove methods took another optional param, that would determine whether the best block number should be updated and how?

It sounds good

Reply:
IMO, in the end you still need to write multiple keys to db, does it matter which part of code handles that?

Reply:
It's about making it clear what the function does. For example currently `removeBlock` would set the best block number to `n - 1`, but what if someone at some point wants to remove a block which is not the best block? He would be surprised, at the least.